cmd/registry-replacer/main.go

```diff
@@ -600,6 +635,88 @@ func pruneReplacements(config *api.ReleaseBuildConfiguration, filter asDirective
 	return utilerrors.NewAggregate(errs)
 }
 
+// when pruning unused base images we'll check the fully-resolved config to make sure an image is not used directly in
+// the config, or within any of the tests.
+func pruneUnusedBaseImages(config *api.ReleaseBuildConfiguration, resolvedConfig *api.ReleaseBuildConfiguration) {
+	usedBaseImages := sets.NewString()
+	for _, step := range resolvedConfig.RawSteps {
+		switch {
+		case step.TestStepConfiguration != nil:
+			getTestStepImages(&usedBaseImages, step.TestStepConfiguration)
+		case step.RPMImageInjectionStepConfiguration != nil:
+			usedBaseImages.Insert(string(step.RPMImageInjectionStepConfiguration.From))
+		}
+	}
+	for _, test := range resolvedConfig.Tests {
+		getTestStepImages(&usedBaseImages, &test)
+	}
+	for _, image := range resolvedConfig.Images {
+		usedBaseImages.Insert(string(image.From))
+		for input := range image.Inputs {
+			usedBaseImages.Insert(input)
+		}
+	}
+
+	for sourceImage := range config.BaseImages {
+		var keep bool
+		for _, candidate := range usedBaseImages.List() {
+			orgRepoTag, _ := orgRepoTagFromPullString(candidate)
+
+			// consider it a match if either the orgRepoTag matches, or the sourceImage matches directly. Depending on
+			// where the image was sourced from it might be in pull string format, or it might just be the image name.
+			keep = orgRepoTag.String() == sourceImage || candidate == sourceImage
+
+			if keep {
+				break
+			}
+		}
+		if !keep {
+			delete(config.BaseImages, sourceImage)
+		}
+	}
+}
+
+func getTestStepImages(images *sets.String, step *api.TestStepConfiguration) {
+	if step.MultiStageTestConfigurationLiteral == nil {
+		return
+	}
+	if len(step.MultiStageTestConfigurationLiteral.Pre) > 0 {
+		getTestImages(images, step.MultiStageTestConfigurationLiteral.Pre)
+	}
+	if len(step.MultiStageTestConfigurationLiteral.Test) > 0 {
+		getTestImages(images, step.MultiStageTestConfigurationLiteral.Test)
+	}
+	if len(step.MultiStageTestConfigurationLiteral.Post) > 0 {
+		getTestImages(images, step.MultiStageTestConfigurationLiteral.Post)
+	}
+}
+
+var pipelineDependencyRegex = regexp.MustCompile(fmt.Sprintf(`(%s:)([\w.\-]+)`, api.PipelineImageStream))
+
+func getTestImages(images *sets.String, steps []api.LiteralTestStep) {
+	for _, step := range steps {
+		if step.From != "" {
+			images.Insert(step.From)
+		}
+
+		// see if we have any pipeline dependency images that also need to be included.
+		if len(step.Dependencies) == 0 {
```

Comment:
Nit: looping over a potentially empty list is fine and there isn't really a need for the extra condition here

Reply:
Right, I keep forgetting that :)